src/CorrelatedNoiseDetection.cc

```diff
@@ -0,0 +1,192 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST/AURA.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "lsst/pex/exceptions.h"
+#include "lsst/afw/math/LeastSquares.h"
+#include "lsst/meas/algorithms/CorrelatedNoiseDetection.h"
+
+namespace lsst { namespace meas { namespace algorithms {
+
+
+afw::image::Image<float> measureCorrelationKernel(
+    afw::image::MaskedImage<float> const & mi,
+    int radius,
+    afw::image::MaskPixel badBitMask
+) {
+    afw::image::Image<float> result(2*radius + 1, 2*radius + 1);
+    afw::image::Image<int> count(result.getDimensions());
+    int const width = mi.getWidth();
+    int const height = mi.getHeight();
+    // iterate over pixels in the MaskedImage, skipping any that meet our bad mask criteria
+    for (int y1 = 0; y1 < height; ++y1) {
+        int const y2a = std::max(0, y1 - radius);
+        int const y2b = std::min(height, y1 + radius + 1);
+        for (int x1 = 0; x1 < width; ++x1) {
+            if ((*mi.getMask())(x1, y1) & badBitMask) {
+                continue;
+            }
+            float z = (*mi.getImage())(x1, y1) / (*mi.getVariance())(x1, y1);
```

Comment:
Don't you need to subtract the mean here and below from the image values?

Reply:
My thinking was that because this should only be running noise pixels, we could assume the mean was zero.  I suppose I should at least add a comment noting that we should check that assumption if we ever use this code in a pipeline setting.

Reply:
I found that there were many regions where this was not the case in the HSC wide data.

Reply:
Of course, they may just get filtered out when taking the median over a tract.

Reply:
Just started thinking about fixing this, and I realized that I don't understand what you're proposing - the "mean" we need to subtract is a correction to the background estimate, right?  How do we estimate that?

Reply:
Yes, depending on how large the image is the background estimate may be off.  Can't you just compute the mean of the image and subtract that off?

Reply:
Yup, if you think just subtracting off the image mean would be an improvement (I was imagining something more local), that's definitely something I can do.